### PR TITLE
feat: Reconciliation in POS using Credit Note

### PR DESCRIPTION
**Pull request by nemesis189**
<h3>Updates:</h3>

- Adding Sales Invoice credit note with outstanding as an advance payment
- Added 'Credit Note' as a Mode of Payment
- Added a checkbox in POS Invoice called `Ignore Payments for Return` to skip payment validation while creating return credit notes via Point Of Sale
---

<details open>
<summary><b>Pre requisite Feats added:</b></summary>
<ol>
<li>Adding Sales Invoice credit note with outstanding as an advance payment </li>
<ul>
<li>Earlier only Payment entires and Journal Entries could be added as an advance for payment in the Advances table.</li>
<li>Now, any sales invoice credit note with outstanding < 0 can be used for advance payment</li>
</ul>

<li>Added 'Credit Note' as a Mode of Payment:</li>
<ul>
<li>A new Mode Of Payment 'Credit Note' has been added which works only for POS currently, it fetches POS Invoice credit notes with negative outstanding</li>
<ul>
</details>

---

<h3>Feature Description:</h3>

- In Point of Sale now we can make use of a POS invoice credit note with negative outstanding for payments. 
- We can select the 'Credit Note' as the mode of payment where all the POS invoice credit notes(returns) with negative outstanding would be listed down and the total of the selected invoice would be added to the `paid_amount` and the outstanding amount of the selected credit note becomes 0.
- While closing a POS, if CN is used as a mode of payment for any invoice, a third consolidated invoice - **Consolidated Advances Invoice** is created along with the usual two other (consolidated sales invoice and consolidated credit return) consolidated invoices.
---
<details open>
<summary><b>Consider the following scenario (also the steps):</b></summary>
<ol>
<li>The first day multiple transactions(invoice) are made by the customer - Make a new opening entry for POS
<li>Create a few pos invoices(3-4) and close the POS
<li>Now the next day one of the invoices needs to be returned so for that - make a new opening entry again(for a new day) and make return for 1-2 invoices through the UI created in the above steps, **while doing so make sure the checkbox `Ignore Payments For Return` is checked(Add it to pos screen via POS Settings). ** This step will make POS return credit notes with negative outstanding which would be used for reconciliation or as payments
<li>Make a new pos sale and in payments select 'Credit Note' as mode of payment and select one of the credit notes created in the above steps.
<li>Complete the transaction, make another transaction with mode of payment as Credit Note.
<li>Close this POS session by creating a closing entry.
<li>Check the 3 consolidated invoices created, the consolidated sales invoice should have the **Consolidated Advances Invoice** added to the **Advances** table of the invoice as an advance payment
</ol>
</details>

- [ ] Add tests

### CODE REVIEW

File: erpnext/accounts/doctype/mode_of_payment/mode_of_payment.json [hunk 1]
```diff
@@ -48,10 +48,12 @@
  "idx": 1,
  "index_web_pages_for_search": 1,
  "links": [],
- "modified": "2020-09-18 17:57:23.835236",
+ "migration_hash": "54043fbbb01893f2866c8367069f2ed0",
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/mode_of_payment/mode_of_payment.json]:
```suggestion
```

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.json [hunk 2]
```diff
@@ -105,12 +106,21 @@
...
+   "label": "Consolidate Advances Credit Note",
+   "options": "Sales Invoice",
+   "read_only": 1
   }
  ],
  "index_web_pages_for_search": 1,
  "is_submittable": 1,
  "links": [],
- "modified": "2021-09-14 11:17:19.001142",
+ "migration_hash": "68e462071d0c4b17941f5e6c57acfd88",
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.json]:
```suggestion
```

File: erpnext/accounts/doctype/pos_settings/pos_settings.json [hunk 3]
```diff
@@ -24,7 +24,8 @@
  ],
  "issingle": 1,
  "links": [],
- "modified": "2021-04-19 14:56:24.465218",
+ "migration_hash": "ec28d79630d0402e6cec3f8b9f061e67",
+ "modified": "2021-11-25 19:59:53.015244",
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_settings/pos_settings.json]:
```suggestion
 "modified": "2021-04-19 14:56:24.465218",
```

File: erpnext/accounts/doctype/sales_invoice_payment/sales_invoice_payment.json [hunk 4]
```diff
@@ -75,11 +76,19 @@
...
+   "fieldname": "credit_note",
+   "fieldtype": "Link",
+   "label": "Credit Note",
+   "options": "POS Invoice"
   }
  ],
  "istable": 1,
  "links": [],
- "modified": "2020-08-03 12:45:39.986598",
+ "migration_hash": "c3a1e585986db6aa4b3cc4f0f24c4400",
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/sales_invoice_payment/sales_invoice_payment.json]:
```suggestion
```

File: erpnext/accounts/doctype/pos_invoice/pos_invoice.py [hunk 5]
```diff
@@ -488,6 +496,30 @@ def get_existing_payment_request(self, pay):
...
+		cn_inv = frappe.get_doc('POS Invoice',cn_mop.credit_note)
+
+		if cn_inv.outstanding_amount :
+			update_cn_outstanding_amount(cn_inv, cn_mop, cancel=cancel)
+
+
+def update_cn_outstanding_amount(cn_inv, cn_mop, cancel=False):
+	cn_inv.flags.ignore_validate_update_after_submit = True
+	if not cancel:
+		cn_inv.outstanding_amount = 0
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice/pos_invoice.py:514]:
```suggestion
		cn_inv.outstanding_amount -= flt(cn_mop.amount)
```

File: erpnext/accounts/doctype/pos_invoice/pos_invoice.py [hunk 6]
```diff
@@ -488,6 +496,30 @@ def get_existing_payment_request(self, pay):
...
+	else:
+		cn_inv.outstanding_amount = flt(cn_mop.amount)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice/pos_invoice.py:516]:
```suggestion
		cn_inv.outstanding_amount += flt(cn_mop.amount)
```

File: erpnext/accounts/doctype/pos_invoice/pos_invoice.py [hunk 7]
```diff
@@ -488,6 +496,30 @@ def get_existing_payment_request(self, pay):
 		if pr:
 			return frappe.get_doc('Payment Request', pr[0][0])
 
+	def update_credit_note(self, cancel=False):
+		# pick out the mop with credit note
+		mop_dict = {p.type:p for p in self.payments}
+		cn_mop = mop_dict.get('Credit Note')
+		if not cn_mop:
+			return
+		cn_inv = frappe.get_doc('POS Invoice',cn_mop.credit_note)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice/pos_invoice.py:505]: There can be multiple Credit Note payments too

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.json [hunk 8]
```diff
@@ -17,6 +17,7 @@
   "consolidated_invoice",
   "column_break_7",
   "consolidated_credit_note",
+  "consolidate_advances_credit_note",
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.json:20]: Rename this to `consolidated_credit_note_payment`

File: erpnext/accounts/doctype/pos_invoice/pos_invoice.json [hunk 9]
```diff
@@ -125,6 +125,7 @@
   "payment_terms_template",
   "payment_schedule",
   "payments_section",
+  "ignore_payments_for_return",
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice/pos_invoice.json:128]: Should be renamed to something like `Use as Payment`
Check with @harshit-30 for a better label

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 10]
```diff
@@ -199,27 +225,89 @@ def merge_pos_invoice_into(self, invoice, data):
...
 			si.cancel()
 
+def get_pos_advance_credit_notes(sales, returns):
+	advances = []
+	cn_amount_mapping = {}
+	for doc in sales:
+		found = 0
+		for p in doc.payments:
+			if found:
+				break
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:285]: There can be multiple Credit Note payments too. So don't `break` just after finding the first one

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 11]
```diff
@@ -199,27 +225,89 @@ def merge_pos_invoice_into(self, invoice, data):
...
+def remove_from_returns_list(returns,doc):
+	for pos_inv in returns:
+		if pos_inv.get('name') == doc:
+			return returns.pop(returns.index(pos_inv))
+
+def set_pos_cn_outstanding_value(advances, cn_amount_mapping, set_zero=False):
+	for cn in advances:
+		cn.load_from_db()
+		cn.flags.ignore_validate_update_after_submit = True
+		cn.outstanding_amount = flt(cn_amount_mapping[cn.name]) if not set_zero else 0.0
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:306]: This should reduce the `outstanding` by the `amount` and incase of cancellation, increase by that value

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 12]
```diff
@@ -268,7 +356,7 @@ def consolidate_pos_invoices(pos_invoices=None, closing_entry=None):
 
 	invoice_by_customer = get_invoice_customer_map(invoices)
 
-	if len(invoices) >= 10 and closing_entry:
+	if len(invoices) >= 100 and closing_entry:
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:359]:
```suggestion
	if len(invoices) >= 10 and closing_entry:
```

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 13]
```diff
@@ -281,7 +369,7 @@ def unconsolidate_pos_invoices(closing_entry):
 		pluck='name'
 	)
 
-	if len(merge_logs) >= 10:
+	if len(merge_logs) >= 100:
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:372]:
```suggestion
	if len(merge_logs) >= 10:
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 14]
```diff
@@ -767,8 +767,7 @@ frappe.ui.form.on('Sales Invoice', {
 
 	hide_fields: function(frm) {
 		let doc = frm.doc;
-		var parent_fields = ['project', 'due_date', 'is_opening', 'source', 'total_advance', 'get_advances',
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js:770]: Use `depends_on` to hide the advance related fields if `set_advances` is not checked or `advances` doesn't have rows

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 15]
```diff
@@ -55,42 +55,61 @@ def validate_pos_invoice_status(self):
...
+		if advances:
+			advance_credit_note = self.process_merging_into_credit_note(advances, is_advance=True)
+			advance_credit_note_doc = frappe.get_doc("Sales Invoice", advance_credit_note)
+
 		if sales:
-			sales_invoice = self.process_merging_into_sales_invoice(sales)
+			sales_invoice = self.process_merging_into_sales_invoice(sales, advance=advance_credit_note_doc)
 
-		self.save() # save consolidated_sales_invoice & consolidated_credit_note ref in merge log
+		self.save() # save consolidated_sales_invoice & consolidated_credit_note ref in merge log)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:73]:
```suggestion
		self.save() # save consolidated_sales_invoice & consolidated_credit_note ref in merge log
```

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 16]
```diff
@@ -55,42 +55,61 @@ def validate_pos_invoice_status(self):
...
 		sales_invoice = self.get_new_sales_invoice()
 
-		sales_invoice = self.merge_pos_invoice_into(sales_invoice, data)
+		sales_invoice = self.merge_pos_invoice_into(sales_invoice, data, advance=advance)
 
 		sales_invoice.is_consolidated = 1
 		sales_invoice.set_posting_time = 1
 		sales_invoice.posting_date = getdate(self.posting_date)
 		sales_invoice.save()
+		# frappe.throw('here')
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:105]:
```suggestion
```

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 17]
```diff
@@ -55,42 +55,61 @@ def validate_pos_invoice_status(self):
...
+		advances, cn_amount_mapping = get_pos_advance_credit_notes(sales, returns)
 
-		sales_invoice, credit_note = "", ""
+		sales_invoice, credit_note, advance_credit_note, advance_credit_note_doc = "", "", "", ""
 		if returns:
 			credit_note = self.process_merging_into_credit_note(returns)
 
+		if advances:
+			advance_credit_note = self.process_merging_into_credit_note(advances, is_advance=True)
+			advance_credit_note_doc = frappe.get_doc("Sales Invoice", advance_credit_note)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:68]: This `get_doc` can be eliminated, if you return `advance_credit_note` doc from `process_merging_into_credit_note`

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 18]
```diff
@@ -199,27 +225,89 @@ def merge_pos_invoice_into(self, invoice, data):
...
 
+
+	def get_advance_pos_credit_notes(self, invoice, payment, advance=None):
+		si = advance
+		advance_row = {
+				"reference_type": "Sales Invoice",
+				"reference_name":si.name,
+				"advance_amount": flt(abs(si.get("outstanding_amount"))),
+				"allocated_amount": flt(payment.get("amount")),
+				# "ref_exchange_rate": flt(si.get("exchange_rate")) # exchange_rate of advance entry
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:236]:
```suggestion
```

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 19]
```diff
@@ -199,27 +225,89 @@ def merge_pos_invoice_into(self, invoice, data):
 
 		return invoice
 
+
+	def get_advance_pos_credit_notes(self, invoice, payment, advance=None):
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:229]: `set_credit_note_payment_as_advance`

File: erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py [hunk 20]
```diff
@@ -199,27 +225,89 @@ def merge_pos_invoice_into(self, invoice, data):
...
+				if doc_is_used_as_cn_mop(doc, invoice_docs):
+					cons_inv = advance_credit_note
+				else:
+					cons_inv = credit_note
+			else:
+				cons_inv = sales_invoice
+
+			if cn_amount_mapping and self.docstatus == 2:
+				# reset the outstanding values on cancellation
+				set_pos_cn_outstanding_value(advances, cn_amount_mapping, set_zero=False)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/pos_invoice_merge_log/pos_invoice_merge_log.py:265]: This should not be done in `update_pos_invoices`

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 21]
```diff
@@ -272,7 +275,8 @@ def validate_pos_return(self):
 				frappe.throw(_("Total payments amount can't be greater than {}").format(-invoice_total))
 
 	def validate_pos_paid_amount(self):
-		if len(self.payments) == 0 and self.is_pos:
+		print('HELLOOOOOO', len(self.payments) == 0 , len(self.advances)==0)
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:278]:
```suggestion
```

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.py [hunk 22]
```diff
@@ -661,7 +665,7 @@ def validate_proj_cust(self):
 				throw(_("Customer {0} does not belong to project {1}").format(self.customer,self.project))
 
 	def validate_pos(self):
-		if self.is_return:
+		if self.is_return and not self.is_pos:
```
Comment (nextchamp-saqib) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.py:668]: This validation was specifically for `is_pos` invoices. This shouldn't be bypassed.

**Review by nextchamp-saqib — CHANGES_REQUESTED** [review]

File: erpnext/controllers/accounts_controller.py [hunk 23]
```diff
@@ -1684,6 +1694,23 @@ def get_advance_payment_entries(party_type, party, party_account, order_doctype,
...
+def get_si_credit_notes(party_type, party, limit=None, condition=None):
+	unallocated_payment_entries = []
+	limit_cond = "limit %s" % limit if limit else ""
+
+	unallocated_payment_entries = frappe.db.sql("""
+			select "Sales Invoice" as reference_type, name as reference_name, posting_date,
+			remarks, abs(outstanding_amount) as amount
+			from `tabSales Invoice`
+			where
+				customer = %s and outstanding_amount != 0
```
Comment (nextchamp-saqib) [inline erpnext/controllers/accounts_controller.py:1706]:
```suggestion
				customer = %s and outstanding_amount > 0
```

**Comment by OneiricIT** [comment]
Any update on this?
Would be finally good to see an easy way to allocate credit on customer account to new / other Sales invoices in system...

https://discuss.erpnext.com/t/allocate-credit-note-debit-note-to-another-sales-invoice-purchase-invoice/29329/6

**Pull request closed without merging**